File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int maxsum=Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        if(root==null)return 0;
        maxgain(root);
        return maxsum;
    }
    public int maxgain(TreeNode root){
        if(root==null)return 0;
        int leftgain=Math.max(maxgain(root.left),0);
        int rightgain=Math.max(maxgain(root.right),0);
        maxsum=Math.max(maxsum,root.val+leftgain+rightgain);
        return root.val+Math.max(leftgain,rightgain);
    }
}

File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int mx=INT_MIN;
    int maxPathSum(TreeNode* root) {
        if(!root)return 0;
        maxgain(root);
        return mx;
    }
    int maxgain(TreeNode* root){
        if(!root)return 0;
        int leftgain=max(maxgain(root->left),0);
        int rightgain=max(maxgain(root->right),0);
        mx=max(mx,leftgain+rightgain+root->val);
        return root->val+max(leftgain,rightgain);
    }
};